lib/drivers/weitai/weiTaiUSB.js

```diff
@@ -0,0 +1,1613 @@
+        insulinSensitivity,
+        bgTarget,
+        basalSchedules,
+        activeName,
+      };
+    },
+
+    buildSettings(settings) {
+      let res = [];
+      //basalSchedules
+      let settingRes = this.buildBasalSchedules(settings);
+      var postsettings = cfg.builder
+        .makePumpSettings()
+        .with_activeSchedule(settingRes.activeName || 'default')
+        .with_units({ carb: 'grams', bg: 'mg/dL' })
+        .with_basalSchedules(settingRes.basalSchedules)
+        .with_carbRatio(settingRes.carbRatio)
+        .with_insulinSensitivity(settingRes.insulinSensitivity)
+        .with_bgTarget(settingRes.bgTarget)
+        .with_manufacturers(['Weitai'])
+        .with_serialNumber(cfg.deviceInfo.serialNumber)
+        .with_deviceTime(sundial.formatDeviceTime(new Date()))
+        .with_time(
+          sundial.applyTimezone(new Date(), cfg.timezone).toISOString()
+        )
+        .with_timezoneOffset(0)
+        .with_conversionOffset(0)
+        .done();
+
+      return postsettings;
+    },
+
+    buildBlood(BloodGlucoses) {
+      let res = [];
+      for (let blood of BloodGlucoses) {
+        if (new Date(blood.deviceTime).valueOf() < cfg.lastUpload) {
+          continue;
+        }
+        const recordBuilder = cfg.builder
+          .makeSMBG()
+          .with_value(parseFloat(blood.BloodGlucose))
+          .with_units('mg/dL') // values are always in 'mg/dL'
+          .with_deviceTime(
+            sundial.formatDeviceTime(new Date(blood.deviceTime).valueOf())
+          )
+          .set('index', blood.recordId);
+
+        cfg.tzoUtil.fillInUTCInfo(
+          recordBuilder,
+          new Date(blood.deviceTime).valueOf()
+        );
+        const postRecord = recordBuilder.done();
+        delete postRecord.index;
+        res.push(postRecord);
+      }
+      return res;
+    },
+
+    buildBasal(BasalRates) {
+      let res = [];
+      //date sort desc
+      BasalRates.sort(function(a, b) {
+        return a.deviceTime < b.deviceTime ? -1 : 1;
+      });
+      //today
+      // BasalRates = BasalRates.filter((a)=>{ return a.deviceTime > '2020-07-20 23:59:59' && a.deviceTime < '2020-07-22 00:00:00'})
+      for (let i = 0; i < BasalRates.length; i++) {
+        let currDu = new Date(BasalRates[i].deviceTime).valueOf();
+        let nextDu = BasalRates[i + 1]
+          ? new Date(BasalRates[i + 1].deviceTime).valueOf()
+          : currDu + 1000;
+        if (nextDu - currDu < 0) {
+          debug('error-basal', BasalRates[i]);
+        }
+        if (i == BasalRates.length - 1) {
+          break;
+        }
+        let basalBuilder = cfg.builder
+          .makeScheduledBasal()
+          .with_scheduleName('currValue')
+          .with_deviceTime(
+            sundial.formatDeviceTime(
+              new Date(BasalRates[i].deviceTime).valueOf()
+            )
+          )
+          // .with_rate(1)
+          .with_rate(BasalRates[i].BasalRate)
+          .with_duration(nextDu - currDu)
```

Comment:
I currently have an issue where I didn't use the pump for a while, and now the duration of the second to last basal is calculated as 1895277000ms (almost 22 days). Does the pump history contain any information on a pump shutdown, so that we can calculate when the basal was not running?